File: Updog.Client/src/user/infrastructure/user-mapper.ts
import { Mapper } from '@/core/mapper';
import { User } from '../domain/user';

/**
 * Mapper to convert a user from it's raw object to an
 * entity.
 */
export class UserMapper implements Mapper<{ [key: string]: any }, User> {
    /**
     * Map a raw object literal to a user.
     * @param source The source object.
     */
    public map(source: { [key: string]: any }): User {
        if (typeof source.id !== 'number') {
            throw new TypeError('id must of type number');
        }

        if (typeof source.username !== 'string') {
            throw new TypeError('username must be of type string');
        }

        if (typeof source.joinedDate !== 'string') {
            throw new TypeError('joinedDate must be a date string');
        }

        return new User(source.id, source.username, new Date(source.joinedDate), source.postKarma, source.commentKarma);
    }
}

File: Updog.Persistance/User/UserMapper.cs
using Updog.Application;
using Updog.Domain;

namespace Updog.Persistance {
    public interface IUserMapper : IReversableMapper<UserRecord, User> { }

    /// <summary>
    /// Mapper to convert the user record into it's user entity.
    /// </summary>
    public sealed class UserMapper : IUserMapper {
        #region Publics
        /// <summary>
        /// Convert the user record into an entity.
        /// </summary>
        /// <param name="source">The record to convert.</param>
        /// <returns>The rebuilt user entity.</returns>
        public User Map(UserRecord source) => new User(
            source.Id,
            source.Username,
            source.Email,
            source.PasswordHash,
            source.JoinedDate,
            source.PostKarma,
            source.CommentKarma
        );

        /// <summary>
        /// Reverse the user entity back into it's record.
        /// </summary>
        /// <param name="destination">The starting entity.</param>
        /// <returns>The rebuilt record.</returns>
        public UserRecord Reverse(User destination) => new UserRecord() {
            Id = destination.Id,
            Username = destination.Username,
            Email = destination.Email,
            PasswordHash = destination.PasswordHash,
            JoinedDate = destination.JoinedDate,
            PostKarma = destination.PostKarma,
            CommentKarma = destination.CommentKarma
        };
        #endregion
    }
}